src/server/cluster/cluster_family.cc

```diff
@@ -626,44 +626,85 @@ static std::string_view state_to_str(MigrationState state) {
   return "UNDEFINED_STATE"sv;
 }
 
-void ClusterFamily::DflyIncomingSlotMigrationStatus(CmdArgList args, ConnectionContext* cntx) {
-  CmdArgParser parser(args);
+void ClusterFamily::SendSingleMigrationStatus(ConnectionContext* cntx, string_view node_id) {
   auto* rb = static_cast<RedisReplyBuilder*>(cntx->reply_builder());
 
+  // find incoming slot migration
+  for (const auto& m : incoming_migrations_jobs_) {
+    if (m->GetSourceID() == node_id)
+      return rb->SendSimpleString(StateToStr(m->GetState()));
+  }
+  // find outgoing slot migration
+  for (const auto& migration : outgoing_migration_jobs_) {
+    if (migration->GetMigrationInfo().node_id == node_id)
+      return rb->SendSimpleString(StateToStr(migration->GetState()));
+  }
+  return rb->SendSimpleString(StateToStr(MigrationState::C_NO_STATE));
+}
+
+void ClusterFamily::SendAllMigrationStatus(ConnectionContext* cntx) {
+  auto* rb = static_cast<RedisReplyBuilder*>(cntx->reply_builder());
+
+  const size_t arr_size = incoming_migrations_jobs_.size() + outgoing_migration_jobs_.size();
+  if (arr_size == 0) {
+    return rb->SendSimpleString(StateToStr(MigrationState::C_NO_STATE));
+  }
+
+  auto get_key_count = [](const SlotRanges& slots) {
+    atomic_uint64_t keys = 0;
+
+    shard_set->pool()->Await([&](auto*) {
+      EngineShard* shard = EngineShard::tlocal();
+      if (shard == nullptr)
+        return;
+
+      uint64_t shard_keys = 0;
+      for (const SlotRange& range : slots) {
+        for (SlotId slot = range.start; slot <= range.end; slot++) {
+          shard_keys += shard->db_slice().GetSlotStats(slot).key_count;
+        }
+      }
+      keys.fetch_add(shard_keys);
+    });
+
+    return keys.load();
+  };
+
+  auto send_answer = [rb](std::string_view direction, std::string_view node_id,
+                          MigrationState state, uint64_t keys) {
+    auto str = absl::StrCat(direction, " ", node_id, " ", StateToStr(state), " ", "key:", keys);
+    rb->SendSimpleString(str);
+  };
+
+  rb->StartArray(arr_size);
```

Comment:
arr_size + 1 because you also send return rb->SendSimpleString(StateToStr(MigrationState::C_NO_STATE));